bundles/org.openhab.binding.bluetooth.bluez/src/main/java/org/openhab/binding/bluetooth/bluez/DBusBlueZBluetoothDevice.java

```diff
@@ -0,0 +1,460 @@
+        }
+    }
+
+    private void onRssiUpdate(RssiEvent event) {
+        this.rssi = (int) event.getRssi();
+        BluetoothScanNotification notification = new BluetoothScanNotification();
+        notification.setRssi(this.rssi);
+        notifyListeners(BluetoothEventType.SCAN_RECORD, notification);
+    }
+
+    private void onConnectedStatusUpdate(ConnectedEvent event) {
+        this.connectionState = event.isConnected() ? ConnectionState.CONNECTED : ConnectionState.DISCONNECTED;
+        notifyListeners(BluetoothEventType.CONNECTION_STATE,
+                new BluetoothConnectionStatusNotification(connectionState));
+    }
+
+    @Override
+    public boolean discoverServices() {
+        com.github.hypfvieh.bluetooth.wrapper.BluetoothDevice dev = device;
+        if (dev == null) {
+            return false;
+        }
+        if (dev.getGattServices().size() > getServices().size()) {
+
+            for (BluetoothGattService dBusBlueZService : dev.getGattServices()) {
+                BluetoothService service = new BluetoothService(UUID.fromString(dBusBlueZService.getUuid()),
+                        dBusBlueZService.isPrimary());
+
+                for (BluetoothGattCharacteristic dBusBlueZCharacteristic : dBusBlueZService.getGattCharacteristics()) {
+                    BluetoothCharacteristic characteristic = new BluetoothCharacteristic(
+                            UUID.fromString(dBusBlueZCharacteristic.getUuid()), 0);
+
+                    for (BluetoothGattDescriptor dBusBlueZDescriptor : dBusBlueZCharacteristic.getGattDescriptors()) {
+                        BluetoothDescriptor descriptor = new BluetoothDescriptor(characteristic,
+                                UUID.fromString(dBusBlueZDescriptor.getUuid()));
+                        characteristic.addDescriptor(descriptor);
+                    }
+
+                    service.addCharacteristic(characteristic);
+                }
+
+                addService(service);
+            }
+
+            notifyListeners(BluetoothEventType.SERVICES_DISCOVERED);
+        }
+        return true;
+    }
+
+    public synchronized void updateDBusBlueZDevice(
+            com.github.hypfvieh.bluetooth.wrapper.BluetoothDevice dBusBlueZDevice) {
+        logger.debug("updateDBusBlueZDevice({})", dBusBlueZDevice);
+        if (this.device != null && this.device == dBusBlueZDevice) {
+            logger.debug("Objects representing the BT device has not changed. Exiting function.");
+            return;
+        }
+
+        this.device = dBusBlueZDevice;
+
+        updateLastSeenTime();
+
+        this.name = dBusBlueZDevice.getName();
+
+        if (Boolean.TRUE.equals(dBusBlueZDevice.isConnected())) {
+            this.connectionState = ConnectionState.CONNECTED;
+        }
+
+        discoverServices();
+    }
+
+    /**
+     * Clean up and release memory.
+     */
+    @Override
+    public void dispose() {
+        this.device = null;
```

Comment:
This is not sufficient cleanup for the bluetooth devices. Dbus will keep a cache of all of the devices it has discovered and you need to make sure that you remove the devices from that cache. The original bluez binding called `remove()` on the underlying device and I suspect you need to do something similar here as well.